File: base-java/src/main/java/tech/thdev/base/view/BasePresenterActivity.java
package tech.thdev.base.view;

import android.os.Bundle;
import android.support.annotation.Nullable;

import tech.thdev.base.presenter.BasePresenter;
import tech.thdev.base.presenter.BaseView;

/**
 * Created by tae-hwan on 5/6/17.
 */

public abstract class BasePresenterActivity<VIEW extends BaseView, P extends BasePresenter<VIEW>> extends BaseActivity implements BaseView {

    protected P presenter;

    protected abstract P onCreatePresenter();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = onCreatePresenter();
        presenter.attachView((VIEW) this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public boolean isNotFinish() {
        return !isFinishing();
    }
}

File: base-kotlin/src/main/java/tech/thdev/base/view/BasePresenterActivity.kt
package tech.thdev.base.view

import android.os.Bundle
import tech.thdev.base.presenter.BasePresenter
import tech.thdev.base.presenter.BaseView

/**
 * Created by Tae-hwan on 9/5/16.
 */
abstract class BasePresenterActivity<in VIEW: BaseView, PRESENTER : BasePresenter<VIEW>> : BaseActivity(), BaseView {

    protected lateinit var presenter: PRESENTER
        private set

    abstract fun onCreatePresenter(): PRESENTER

    @Suppress("UNCHECKED_CAST")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = onCreatePresenter()
        presenter.attachView(this as VIEW)
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.detachView()
    }
}
